File: src/converter/plugins/InheritanceModifierPlugin.ts
import {ConverterPlugin} from "../plugin.js";
import ts, {Node} from "typescript";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {InheritanceModifier, InheritanceModifierContext} from "../inheritanceModifier.js";
import {GeneratedFile} from "../generated.js";
import {Signature} from "./convertParameterDeclaration.js";

export const inheritanceModifierServiceKey = Symbol()

export class InheritanceModifierService {
    constructor(private readonly inheritanceModifiers: InheritanceModifier[]) {
    }

    resolveSignatureInheritanceModifier(
        node: Node,
        signature: Signature,
        context: ConverterContext,
    ): string | null {
        const inheritanceModifierContext: InheritanceModifierContext = {
            ...context,
            signature,
        }

        return this.internalResolveInheritanceModifier(node, inheritanceModifierContext)
    }

    resolveGetterInheritanceModifier(
        node: Node,
        context: ConverterContext,
    ): string | null {
        const inheritanceModifierContext: InheritanceModifierContext = {
            ...context,
            getter: true,
        }

        return this.internalResolveInheritanceModifier(node, inheritanceModifierContext)
    }

    resolveSetterInheritanceModifier(
        node: Node,
        context: ConverterContext,
    ): string | null {
        const inheritanceModifierContext: InheritanceModifierContext = {
            ...context,
            setter: true,
        }

        return this.internalResolveInheritanceModifier(node, inheritanceModifierContext)
    }

    resolveInheritanceModifier(
        node: Node,
        context: ConverterContext,
    ): string | null {
        return this.internalResolveInheritanceModifier(node, context)
    }

    private internalResolveInheritanceModifier(
        node: Node,
        context: InheritanceModifierContext
    ): string | null {
        for (const inheritanceModifier of this.inheritanceModifiers) {
            const result = inheritanceModifier(node, context)

            if (result !== null) return result
        }

        return null
    }
}

export class InheritanceModifierPlugin implements ConverterPlugin {
    private readonly inheritanceModifierService: InheritanceModifierService;

    constructor(inheritanceModifiers: InheritanceModifier[]) {
        this.inheritanceModifierService = new InheritanceModifierService(inheritanceModifiers)
    }

    generate(): GeneratedFile[] {
        return [];
    }

    render(node: Node, context: ConverterContext, next: Render): string | null {
        return null;
    }

    traverse(node: ts.Node, context: ConverterContext): void {
    }

    setup(context: ConverterContext): void {
        context.registerService(inheritanceModifierServiceKey, this.inheritanceModifierService)
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/InheritanceModifierPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.*
import js.array.ReadonlyArray
import js.symbol.Symbol
import typescript.Node

@OptIn(ExperimentalJsExport::class)
@JsExport
val inheritanceModifierServiceKey = Symbol()

class InheritanceModifierService(private val inheritanceModifiers: ReadonlyArray<InheritanceModifier<Node>>) {
    fun resolveSignatureInheritanceModifier(
        node: Node,
        signature: Signature,
        context: ConverterContext,
    ): String? {
        val inheritanceModifierContext = object : InheritanceModifierContext, ConverterContext by context {
            override val signature = signature
            override val getter = false
            override val setter = false
        }

        return internalResolveInheritanceModifier(node, inheritanceModifierContext)
    }

    fun resolveGetterInheritanceModifier(
        node: Node,
        context: ConverterContext,
    ): String? {
        val inheritanceModifierContext = object : InheritanceModifierContext, ConverterContext by context {
            override val signature = null
            override val getter = true
            override val setter = false
        }

        return internalResolveInheritanceModifier(node, inheritanceModifierContext)
    }

    fun resolveSetterInheritanceModifier(
        node: Node,
        context: ConverterContext,
    ): String? {
        val inheritanceModifierContext = object : InheritanceModifierContext, ConverterContext by context {
            override val signature = null
            override val getter = false
            override val setter = true
        }

        return internalResolveInheritanceModifier(node, inheritanceModifierContext)
    }

    fun resolveInheritanceModifier(
        node: Node,
        context: ConverterContext,
    ): String? {
        val inheritanceModifierContext = object : InheritanceModifierContext, ConverterContext by context {
            override val signature = null
            override val getter = false
            override val setter = false
        }

        return internalResolveInheritanceModifier(node, inheritanceModifierContext)
    }

    private fun internalResolveInheritanceModifier(
        node: Node,
        context: InheritanceModifierContext
    ): String? {
        for (inheritanceModifier in inheritanceModifiers) {
            val result = inheritanceModifier(node, context)

            if (result != null) return result
        }

        return null
    }
}

class InheritanceModifierPlugin(inheritanceModifiers: ReadonlyArray<InheritanceModifier<Node>>) : ConverterPlugin<Node> {
    private val inheritanceModifierService = InheritanceModifierService(inheritanceModifiers)

    override fun setup(context: ConverterContext) {
        context.registerService(inheritanceModifierServiceKey, this.inheritanceModifierService)
    }

    override fun traverse(node: Node, context: ConverterContext) = Unit

    override fun render(node: Node, context: ConverterContext, next: Render<Node>) = null

    override fun generate(context: ConverterContext, render: Render<Node>) = emptyArray<GeneratedFile>()
}
